mapeditor/inspector/townevent.cpp

```diff
@@ -0,0 +1,320 @@
+{
+	QVariantMap descriptor;
+	descriptor["name"] = ui->eventNameText->text();
+	descriptor["message"] = ui->eventMessageText->toPlainText();
+	descriptor["humanAffected"] = QVariant::fromValue(ui->eventAffectsHuman->isChecked());
+	descriptor["computerAffected"] = QVariant::fromValue(ui->eventAffectsCpu->isChecked());
+	descriptor["firstOccurrence"] = QVariant::fromValue(ui->eventFirstOccurrence->value()-1);
+	descriptor["nextOccurrence"] = QVariant::fromValue(ui->eventRepeatAfter->value());
+	descriptor["players"] = playersToVariant();
+	descriptor["resources"] = resourcesToVariant();
+	descriptor["buildings"] = buildingsToVariant();
+	descriptor["creatures"] = creaturesToVariant();
+
+	item->setData(Qt::UserRole, descriptor);
+	auto itemText = tr("Day %1 - %2").arg(ui->eventFirstOccurrence->value(), 3).arg(ui->eventNameText->text());
+	item->setText(itemText);
+}
+
+QVariant TownEvent::playersToVariant()
+{
+	int players = 0;
+	for (int i = 0; i < ui->playersAffected->count(); ++i)
+	{
+		auto * item = ui->playersAffected->item(i);
+		if (item->checkState() == Qt::Checked)
+			players |= 1 << i;
+	}
+	return QVariant::fromValue(players);
+}
+
+QVariantMap TownEvent::resourcesToVariant()
+{
+	auto res = item->data(Qt::UserRole).toMap().value("resources").toMap();
+	for (int i = 0; i < GameConstants::RESOURCE_QUANTITY; ++i)
+	{
+		auto * itemType = ui->resourcesTable->item(i, 0);
+		auto * itemQty = static_cast<QSpinBox *> (ui->resourcesTable->cellWidget(i, 1));
+
+		res[itemType->text()] = QVariant::fromValue(itemQty->value());
+	}
+	return res;
+}
+
+QVariantList TownEvent::buildingsToVariant()
+{
+	QVariantList buildingsList;
+	std::vector<QModelIndex> stack;
+	stack.push_back(QModelIndex());
+	while (!stack.empty())
+	{
+		auto pindex = stack.back();
+		stack.pop_back();
+		for (int i = 0; i < buildingsModel.rowCount(pindex); ++i)
+		{
+			QModelIndex index = buildingsModel.index(i, 1, pindex);
+			if (auto * item = buildingsModel.itemFromIndex(index))
+				if (item->checkState() == Qt::Checked)
+					buildingsList.push_back(item->data(Qt::UserRole));
+			index = buildingsModel.index(i, 0, pindex);
+			if (buildingsModel.hasChildren(index))
+				stack.push_back(index);
+		}
+	}
+	return buildingsList;
+}
+
+QVariantList TownEvent::creaturesToVariant()
+{
+	QVariantList creaturesList;
+	for (int i = 0; i < 7; ++i)
+	{
+		auto * item = static_cast<QSpinBox *>(ui->creaturesTable->cellWidget(i, 1));
+		creaturesList.push_back(item->value());
+	}
+	return creaturesList;
+}
+
+void TownEvent::on_okButton_clicked()
+{
+	close();
+}
+
+void TownEvent::setRowColumnCheckState(QStandardItem * item, int column, Qt::CheckState checkState) {
+	auto sibling = item->model()->sibling(item->row(), column, item->index());
+	buildingsModel.itemFromIndex(sibling)->setCheckState(checkState);
+}
+
+void TownEvent::onItemChanged(QStandardItem * item)
+{
+	disconnect(&buildingsModel, &QStandardItemModel::itemChanged, this, &TownEvent::onItemChanged);
```

Comment:
I think you might want to use QSignalBlocker class here instead

Reply:
tried and works, but have side effect of not updating checkboxes until user hovers over them or move mouse outside of buildings treeview, so I guess I will live it as is. A little bit uglier code for better user experience